vulkan/vulkan_application.cc

```diff
@@ -0,0 +1,156 @@
+
+VulkanApplication::VulkanApplication(
+    VulkanProcTable& p_vk,
+    const std::string& application_name,
+    std::vector<std::string> enabled_extensions,
+    uint32_t application_version,
+    uint32_t api_version)
+    : vk(p_vk), api_version_(api_version), valid_(false) {
+  // Check if we want to enable debugging.
+
+  bool enable_instance_debugging =
+      EnableDebugging() && VulkanDebugReport::DebugExtensionSupported(vk);
+
+  // Configure extensions.
+
+  if (enable_instance_debugging) {
+    enabled_extensions.emplace_back(VulkanDebugReport::DebugExtensionName());
+  }
+
+  const char* extensions[enabled_extensions.size()];
+
+  for (size_t i = 0; i < enabled_extensions.size(); i++) {
+    extensions[i] = enabled_extensions[i].c_str();
+  }
+
+  // Configure layers.
+
+  const std::vector<std::string> enabled_layers = InstanceLayersToEnable(vk);
+
+  const char* layers[enabled_layers.size()];
+
+  for (size_t i = 0; i < enabled_layers.size(); i++) {
+    layers[i] = enabled_layers[i].c_str();
+  }
+
+  // Configure init structs.
+
+  const VkApplicationInfo info = {
+      .sType = VK_STRUCTURE_TYPE_APPLICATION_INFO,
+      .pNext = nullptr,
+      .pApplicationName = application_name.c_str(),
+      .applicationVersion = application_version,
+      .pEngineName = "FlutterEngine",
+      .engineVersion = VK_MAKE_VERSION(1, 0, 0),
+      .apiVersion = api_version_,
+  };
+
+  const VkInstanceCreateInfo create_info = {
+      .sType = VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO,
+      .pNext = nullptr,
+      .flags = 0,
+      .pApplicationInfo = &info,
+      .enabledLayerCount = static_cast<uint32_t>(enabled_layers.size()),
+      .ppEnabledLayerNames = layers,
+      .enabledExtensionCount = static_cast<uint32_t>(enabled_extensions.size()),
+      .ppEnabledExtensionNames = extensions,
+  };
+
+  // Perform initialization.
+
+  VkInstance instance = VK_NULL_HANDLE;
+
+  if (VK_CALL_LOG_ERROR(vk.CreateInstance(&create_info, nullptr, &instance)) !=
+      VK_SUCCESS) {
+    return;
+  }
+
+  // Now that we have an instance, setup instance proc table entries.
+  if (!vk.SetupInstanceProcAddresses(instance)) {
+    return;
+  }
+
+  instance_ = {instance, [this](VkInstance i) {
+                 FTL_LOG(INFO) << "Destroying instance";
+                 vk.DestroyInstance(i, nullptr);
+               }};
+
+  if (enable_instance_debugging) {
+    auto debug_report = std::make_unique<VulkanDebugReport>(vk, instance_);
+    if (!debug_report->IsValid()) {
+      FTL_LOG(INFO) << "Vulkan debugging was enabled but could not be setup "
+                       "for this instance.";
+    } else {
+      debug_report_ = std::move(debug_report);
+    }
+  }
+
+  valid_ = true;
+}
+
+VulkanApplication::~VulkanApplication() = default;
+
+bool VulkanApplication::IsValid() const {
+  return valid_;
+}
+
+uint32_t VulkanApplication::APIVersion() const {
+  return api_version_;
+}
+
+const VulkanHandle<VkInstance>& VulkanApplication::Instance() {
```

Comment:
`GetInstance` is the conventions. Brain damage from writing too much [Objective-C](https://developer.apple.com/library/content/documentation/Cocoa/Conceptual/ProgrammingWithObjectiveC/Conventions/Conventions.html). Fixing.